### PR TITLE
3076/How_to_change_languages

### LINKED ISSUE #3076: How Change Languages
i modified 
/opt/privacyidea/lib/python3.8/site-packages/privacyidea/app.py
image
/opt/privacyidea/lib/python3.8/site-packages/privacyidea/webui/login.py
image
but I have also tried with several browsers, in all the language settings are zh_Hant and in all of them the pages are shown in English

### PR DESCRIPTION
**Pull request by NNLemling**
The option to define a list of preferred languages ​​in the pi.cfg.

closes #3076

### CODE REVIEW

File: privacyidea/webui/login.py [hunk 1]
```diff
@@ -46,6 +47,22 @@
...
+
+def get_accepted_language(req):
+    # if we are not in the request context, return None to use the default
+    # locale
+    if not req:
+        return None
+    # otherwise try to guess the language from the user accept
+    # header the browser transmits. The best match wins.
+    pi_cfg_configuration = get_app_config_value("PREFERRED_LANGUAGE")
+    # reads pi.cfg and checks if preferred language is set. Otherwise default list ist selected.
```
[the remark this answers is not in the document]
Comment (jona-samuel) [reply privacyidea/webui/login.py:62]: please correct this to is

**Review by jona-samuel — CHANGES_REQUESTED** [review]
the only thing to change is the one typo

File: privacyidea/webui/login.py [hunk 2]
```diff
@@ -45,10 +45,10 @@
 from privacyidea.lib.queue import has_job_queue
 
 DEFAULT_THEME = "/static/contrib/css/bootstrap-theme.css"
+DEFAULT_LANGUAGE_LIST = ["en", "de", "de_DE", "nl", "fr",
+                         "cs", "es", "tr", "zh_Hant"]
```
Comment (plettich) [inline privacyidea/webui/login.py:49]: According to our weblate translations the languages with more than 50% translations are:
`['en', 'de', 'nl', 'zh_Hant', 'fr', 'es', 'tr']`

File: doc/installation/system/inifile.rst [hunk 3]
```diff
@@ -112,6 +112,16 @@ If ``PI_HSM_PADDING`` is not set, the default internal padding is used::
...
 
+``PREFERRED_LANGUAGE`` is a list in which the preferred languages can be defined. The browser's language settings are compared to this
+list and the "best match" wins. If none of the languages set in the browser match, the first language in the list will be used as the output language::
+
+    PREFERRED_LANGUAGE = ["en", "de", "es", "fr"]
+
+.. note:: If ``PREFERRED_LANGUAGE`` is not defined, a ``DEFAULT_LANGUAGE_LIST`` is used.
+   DEFAULT_LANGUAGE_LIST = ["en", "de", "de_DE", "nl", "fr", "cs", "es", "tr", "zh_Hant"]
+
+   If none of these languages are defined in the browser, the first language in the list will also be used here.
```
Comment (plettich) [inline doc/installation/system/inifile.rst:123]: I think this should be clear from the text above.

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/webui/login.py (hunk 2 above)
Comment (NNLemling) [reply privacyidea/webui/login.py]: fixed in 09d11230e87bdf58510ca5d85aed0f8271538b71.

File: doc/installation/system/inifile.rst [hunk 4]
```diff
@@ -112,6 +112,16 @@ If ``PI_HSM_PADDING`` is not set, the default internal padding is used::
 .. warning:: The two paddings are not compatible. Data encrypted with one padding can not decrypted with the other padding. 
    So if you change the padding later during the cause of operation, it will not be possible to decrypt old data!
 
+``PREFERRED_LANGUAGE`` is a list in which the preferred languages can be defined. The browser's language settings are compared to this
+list and the "best match" wins. If none of the languages set in the browser match, the first language in the list will be used as the output language::
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/installation/system/inifile.rst]: fixed in 09d11230e87bdf58510ca5d85aed0f8271538b71.

File: doc/installation/system/inifile.rst (hunk 3 above)
Comment (NNLemling) [reply doc/installation/system/inifile.rst]: fixed in 09d11230e87bdf58510ca5d85aed0f8271538b71.

File: privacyidea/webui/login.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NNLemling) [reply privacyidea/webui/login.py]: fixed in 4cc293e9736c263ae3c8c3cde015e5f43cbfbe5b.

**Review by jona-samuel — APPROVED** [review]

File: privacyidea/webui/login.py [hunk 5]
```diff
@@ -44,10 +45,22 @@
...
 
+def get_accepted_language(req):
+    # if we are not in the request context, return None to use the default
+    # locale
+    if not req:
+        return None
+    # otherwise try to guess the language from the user accept
+    # header the browser transmits. The best match wins.
+    pi_cfg_configuration = get_app_config_value("PREFERRED_LANGUAGE", default=DEFAULT_LANGUAGE_LIST)
+    # reads pi.cfg and checks if preferred language is set. Otherwise default list is selected.
```
Comment (plettich) [inline privacyidea/webui/login.py:61]: i think we need to switch the comments here with L58-59
Also the variable name is misleading, just use something like `pi_lang_list`

File: doc/installation/system/inifile.rst [hunk 6]
```diff
@@ -112,6 +112,14 @@ If ``PI_HSM_PADDING`` is not set, the default internal padding is used::
 .. warning:: The two paddings are not compatible. Data encrypted with one padding can not decrypted with the other padding. 
    So if you change the padding later during the cause of operation, it will not be possible to decrypt old data!
 
+``PREFERRED_LANGUAGE`` is a list in which the preferred languages can be defined. The browser's language settings are compared to this
```
Comment (plettich) [inline doc/installation/system/inifile.rst:115]: As discussed, this should be something like `PI_PREFERRED_LANGUAGE`

File: doc/installation/system/inifile.rst [hunk 7]
```diff
@@ -112,6 +112,14 @@ If ``PI_HSM_PADDING`` is not set, the default internal padding is used::
...
+list and the "best match" wins. If none of the languages set in the browser match, the first language in the list will be used as the default language::
+
+    PREFERRED_LANGUAGE = ["en", "de", "es", "fr"]
+
+.. note:: If ``PREFERRED_LANGUAGE`` is not defined, a ``DEFAULT_LANGUAGE_LIST`` is used.
+   DEFAULT_LANGUAGE_LIST = ['en', 'de', 'nl', 'zh_Hant', 'fr', 'es', 'tr']
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/installation/system/inifile.rst]: fixed in f1975862ca4f27f3f4a5882c97848bc68a575245.

File: privacyidea/webui/login.py (hunk 5 above)
Comment (NNLemling) [reply privacyidea/webui/login.py:58]: fixed in f1975862ca4f27f3f4a5882c97848bc68a575245.

File: doc/installation/system/inifile.rst (hunk 6 above)
Comment (NNLemling) [reply doc/installation/system/inifile.rst]: fixed in f1975862ca4f27f3f4a5882c97848bc68a575245.

File: privacyidea/webui/login.py [hunk 8]
```diff
@@ -44,10 +45,22 @@
...
 
 
+def get_accepted_language(req):
+    # if we are not in the request context, return None to use the default
+    # locale
+    if not req:
+        return None
+    # reads pi.cfg and checks if preferred language is set. Otherwise default list is selected.
+    pi_lang_list = get_app_config_value("PI_PREFERRED_LANGUAGE", default=DEFAULT_LANGUAGE_LIST)
+    # otherwise try to match the language from the user
```
Comment (plettich) [inline privacyidea/webui/login.py:60]: Now the `otherwise` is misleading :-)

File: doc/installation/system/inifile.rst [hunk 9]
```diff
@@ -112,6 +112,14 @@ If ``PI_HSM_PADDING`` is not set, the default internal padding is used::
 .. warning:: The two paddings are not compatible. Data encrypted with one padding can not decrypted with the other padding. 
    So if you change the padding later during the cause of operation, it will not be possible to decrypt old data!
 
+``PI_PREFERRED_LANGUAGE`` is a list in which the preferred languages can be defined. The browser's language settings are compared to this
+list and the "best match" wins. If none of the languages set in the browser match, the first language in the list will be used as the default language::
+
+    PI_PREFERRED_LANGUAGE = ["en", "de", "es", "fr"]
+
+.. note:: If ``PI_PREFERRED_LANGUAGE`` is not defined, a ``DEFAULT_LANGUAGE_LIST`` is used.
```
Comment (plettich) [inline doc/installation/system/inifile.rst:120]: just remove the `DEFAULT_LANGUAGE_LIST` and replace it with something like "... the following list is used"

**Review by plettich — CHANGES_REQUESTED** [review]

File: doc/installation/system/inifile.rst (hunk 9 above)
Comment (NNLemling) [reply doc/installation/system/inifile.rst]: fixed in be6d9a01cd35589755d7885117001b48ceaca91c.

**Review by plettich — APPROVED** [review]

**Pull request merged**